src/firmware/shared/circular_buffer.c

```diff
@@ -0,0 +1,68 @@
+#include "firmware/shared/circular_buffer.h"
+
+#include <assert.h>
+
+struct CircularBuffer
+{
+    int head;
+    int tail;
+    size_t max_size;
+    float *buffer_data;
+    bool isFull;
+};
+
+CircularBuffer_t *circular_buffer_create(size_t size)
+{
+    CircularBuffer_t *new_cbuffer = malloc(sizeof(CircularBuffer_t));
+
+    new_cbuffer->head        = 0;
+    new_cbuffer->tail        = 0;
+    new_cbuffer->max_size    = size;
+    new_cbuffer->buffer_data = (float *)malloc(sizeof(float) * new_cbuffer->max_size);
+    new_cbuffer->isFull      = false;
+
+    return new_cbuffer;
+}
+
+void circular_buffer_destroy(CircularBuffer_t *cbuffer)
+{
+    free(cbuffer);
+}
+
+void circular_buffer_push(CircularBuffer_t *cbuffer, float data)
+{
+    cbuffer->buffer_data[cbuffer->head] = data;
+    cbuffer->head                       = (++cbuffer->head) % cbuffer->max_size;
+    if (circular_buffer_is_full(cbuffer) == true)
+    {
+        cbuffer->tail = (++cbuffer->tail) % cbuffer->max_size;
+    }
+
+    // Implementation assumes head and tail are equal when full
```

Comment:
nit / suggestion: You could technically use this fact for the implementation of `is_full`, and then you wouldn't need to keep track of the `is_full` boolean. You'd have to change the initialization of the struct so the head/tail aren't equal at the start but I think that's valid? This class is simple though so not super necessary.

Reply:
I'll keep that in mind for the future.

Reply:
I don't see an easy way to remove is_full, so I'm fine with leaving it as is